exchanges/poloniex/poloniex_futures.go

```diff
@@ -0,0 +1,741 @@
+	}
+	var resp *FuturesMarkPrice
+	return resp, e.SendHTTPRequest(ctx, exchange.RestSpot, request.UnAuth, common.EncodeURLValues(marketsPathV3+"markPrice", params), &resp)
+}
+
+// GetMarkPriceKlineData obtain the K-line data for the mark price.
+func (e *Exchange) GetMarkPriceKlineData(ctx context.Context, symbol string, interval kline.Interval, startTime, endTime time.Time, limit uint64) ([]*FuturesMarkPriceCandle, error) {
+	if symbol == "" {
+		return nil, currency.ErrSymbolStringEmpty
+	}
+	intervalString, err := intervalToString(interval)
+	if err != nil {
+		return nil, err
+	}
+	params := url.Values{}
+	params.Set("symbol", symbol)
+	params.Set("interval", intervalString)
+	if !startTime.IsZero() && !endTime.IsZero() {
+		if err := common.StartEndTimeCheck(startTime, endTime); err != nil {
+			return nil, err
+		}
+		params.Set("sTime", strconv.FormatInt(startTime.UnixMilli(), 10))
+		params.Set("eTime", strconv.FormatInt(startTime.UnixMilli(), 10))
+	}
+	if limit > 0 {
+		params.Set("limit", strconv.FormatUint(limit, 10))
+	}
+	var resp []*FuturesMarkPriceCandle
+	return resp, e.SendHTTPRequest(ctx, exchange.RestSpot, request.UnAuth, common.EncodeURLValues(marketsPathV3+"markPriceCandlesticks", params), &resp)
+}
+
+// GetFuturesAllProducts inquire about the basic information of the all product.
+func (e *Exchange) GetFuturesAllProducts(ctx context.Context, symbol string) ([]*ProductInfo, error) {
+	params := url.Values{}
+	if symbol != "" {
+		params.Set("symbol", symbol)
+	}
+	var resp []*ProductInfo
+	return resp, e.SendHTTPRequest(ctx, exchange.RestSpot, request.UnAuth, common.EncodeURLValues(marketsPathV3+"allInstruments", params), &resp)
+}
+
+// GetFuturesProduct inquire about the basic information of the product.
+func (e *Exchange) GetFuturesProduct(ctx context.Context, symbol string) (*ProductInfo, error) {
+	if symbol == "" {
+		return nil, currency.ErrSymbolStringEmpty
+	}
+	params := url.Values{}
+	params.Set("symbol", symbol)
+	var resp *ProductInfo
+	return resp, e.SendHTTPRequest(ctx, exchange.RestSpot, request.UnAuth, common.EncodeURLValues(marketsPathV3+"instruments", params), &resp)
+}
+
+// GetFuturesCurrentFundingRate retrieve the current funding rate of the contract.
+func (e *Exchange) GetFuturesCurrentFundingRate(ctx context.Context, symbol string) (*FuturesFundingRate, error) {
+	if symbol == "" {
+		return nil, currency.ErrSymbolStringEmpty
+	}
+	params := url.Values{}
+	params.Set("symbol", symbol)
+	var resp *FuturesFundingRate
+	return resp, e.SendHTTPRequest(ctx, exchange.RestSpot, request.UnAuth, common.EncodeURLValues(marketsPathV3+"fundingRate", params), &resp)
+}
+
+// GetFuturesHistoricalFundingRates retrieve the previous funding rates of a contract.
+func (e *Exchange) GetFuturesHistoricalFundingRates(ctx context.Context, symbol string, startTime, endTime time.Time, limit uint64) ([]*FuturesFundingRate, error) {
+	params := url.Values{}
+	if symbol != "" {
+		params.Set("symbol", symbol)
```

Comment:
The tests is passing but it should fail:

```
Poloniex attempt 1 request path: https://api.poloniex.com/v3/market/fundingRate/history?limit=100
Poloniex request type: GET
HTTP status: 200 OK, Code: 200
Poloniex raw response: {"code":400,"msg":"Param error symbol","data":[]}
--- PASS: TestGetFuturesHistoricalFundingRates (0.18s)
PASS
```

Reply:
@samuael You haven't fixed the underlying problem here which allows this test to pass 

it returns `Poloniex raw response: {"code":400,"msg":"Param error symbol","data":[]}`

This will obfuscate potential issues in the future; please handle code != 200 for unauth and auth requests.

Reply:
I don't know why it is not showing up in your side but I have resolved the issue.

<img width="818" height="185" alt="Screenshot 2025-10-28 at 9 54 39 at night" src="https://github.com/user-attachments/assets/088913a9-c475-41e7-85e9-806c12f21f84" />
<img width="946" height="185" alt="Screenshot 2025-10-28 at 9 54 55 at night" src="https://github.com/user-attachments/assets/fe10c9fe-898d-4531-9e24-8d003257eb97" />

Reply:
I did test this, you have resolved the top level issue, but the core issue I was pointing at was the fact that when it 400's the error is not passed back to the caller. If you remove the empty symbol check and you pass an empty string it will error but the test will pass. This must not occur. 

<img width="662" height="116" alt="image" src="https://github.com/user-attachments/assets/f46a6ad7-70f2-409f-814a-19130deb146f" />

If you have fixed it up cool 👍 but I will be retesting this today. 🚀